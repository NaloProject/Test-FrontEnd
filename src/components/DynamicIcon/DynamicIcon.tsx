import React, { useState, useEffect, useRef, FC } from 'react'
import { ReactComponent as PlaceHolderIcon } from '@Svg/placeholder.svg'

export type DynamicIconProps = {
  className?: string
  name: string
  size?: number | string
  fill?: string
}

const DynamicIcon: FC<DynamicIconProps> = ({
  name,
  size = 24,
  fill,
  ...rest
}) => {
  const ImportedIconRef = useRef<LazyIcon>(null)
  const [loading, setLoading] = useState(false)

  useEffect(() => {
    setLoading(true)
    const importIcon = async () => {
      try {
        const { ReactComponent: namedImport } = await import(
          `../../svg/${name}.svg`
        )
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        ImportedIconRef.current = namedImport
        // eslint-disable-next-line no-useless-catch
      } catch (err) {
        throw err
      } finally {
        setLoading(false)
      }
    }
    importIcon()
  }, [name])

  if (loading) {
    return <PlaceHolderIcon width={size} height={size} fill={fill} {...rest} />
  }

  if (!loading && ImportedIconRef.current) {
    const { current: ImportedIcon } = ImportedIconRef
    return <ImportedIcon width={size} height={size} fill={fill} {...rest} />
  }

  return null
}

export default DynamicIcon
