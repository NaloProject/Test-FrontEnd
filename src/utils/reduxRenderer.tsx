import { configureStore, EnhancedStore } from '@reduxjs/toolkit'
import { render, RenderOptions, RenderResult } from '@testing-library/react'
import { ReactElement, FC, PropsWithChildren } from 'react'
import { Provider } from 'react-redux'

import { combined, RootState } from '@Redux/store'
import INITIAL_STATE from './initialState'

type RenderWithReduxOptions = Omit<RenderOptions, 'queries'> & {
  initialState?: RootState
  store?: EnhancedStore
}

export const renderWithRedux = (
  ui: ReactElement,
  {
    initialState = INITIAL_STATE,
    store = configureStore({
      middleware: (getDefaultMiddleware) =>
        getDefaultMiddleware({
          immutableCheck: false,
          serializableCheck: false,
        }),
      preloadedState: initialState,
      reducer: combined,
    }),
    ...renderOptions
  }: RenderWithReduxOptions = {}
): RenderResult => {
  const Wrapper: FC<PropsWithChildren> = ({ children }) => (
    <Provider store={store}>{children}</Provider>
  )

  return render(ui, { wrapper: Wrapper, ...renderOptions })
}
