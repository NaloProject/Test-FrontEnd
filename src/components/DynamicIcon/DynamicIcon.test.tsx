import { renderWithRedux } from '@Utils/reduxRenderer'
import DynamicIcon from './DynamicIcon'

describe('DynamicIcon', () => {
  it('should mount first', async () => {
    const { container } = renderWithRedux(<DynamicIcon size={32} name="moon" />)

    expect(container).toMatchSnapshot()
  })
})
