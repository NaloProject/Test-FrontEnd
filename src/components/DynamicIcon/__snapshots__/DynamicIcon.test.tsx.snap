// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicIcon should mount first 1`] = `
<div>
  <svg
    height="32"
    width="32"
  />
</div>
`;
